rs/index/src/ivf/writer.rs

```diff
@@ -183,14 +184,22 @@ impl<Q: Quantizer, C: IntSeqEncoder + 'static> IvfWriter<Q, C> {
         metadata_bytes_written +=
             wrap_write(&mut metadata_writer, &num_posting_lists.to_le_bytes())?;
         for i in 0..num_posting_lists {
-            let posting_list = ivf_builder
-                .posting_lists()
-                .get(i as u32)?
-                .iter()
-                .collect::<Vec<_>>();
-            let mut encoder = C::new_encoder(/*universe*/ None, posting_list.len());
-            // Encode to get the length of the encoded data
-            encoder.encode(&posting_list)?;
+            let posting_list = ivf_builder.posting_lists().get(i as u32)?;
+            let encoder = if posting_list.slices.len() == 1 {
+                // If the posting list is contiguous, we can directly create the encoder
+                // without collecting into a Vec
+                let slice = transmute_u8_to_slice::<u64>(posting_list.slices[0]);
+                let len = slice.len() / std::mem::size_of::<u64>();
+                let mut encoder = C::new_encoder(/*universe*/ None, len);
+                encoder.encode(slice)?;
+                encoder
+            } else {
+                // If it's not contiguous, we still need to collect
+                let collected_list = posting_list.iter().collect::<Vec<_>>();
+                let mut encoder = C::new_encoder(/*universe*/ None, collected_list.len());
+                encoder.encode(&collected_list)?;
+                encoder
```

Comment:
this is such wasted work - we collect them into a vec, then inside there, we will iterate again. I think a better way to do is to just change the API of `encode` to take in an iterator (lifetime might be a bit complicated). 

it's ok for now - i'll optimize when I do another round of optimization

Reply:
I wonder if you can just iterate through the slice and repeatedly call `encode`?

so the `encode` function can just append more things into the internal representation?

Reply:
> this is such wasted work - we collect them into a vec, then inside there, we will iterate again. I think a better way to do is to just change the API of encode to take in an iterator (lifetime might be a bit complicated).

Yes, I know, that's why I'm trying to optimize it here. The problem is for the case where the posting list does not fit into a single mmap page, we'll need to read the whole posting list into a temporary buffer.

> I wonder if you can just iterate through the slice and repeatedly call encode?

Makes sense, I think this could work